compaction/compaction_manager.cc

```diff
@@ -1185,6 +1185,33 @@ future<> compaction_manager::perform_sstable_scrub_validate_mode(replica::table*
     return perform_task(seastar::make_shared<validate_sstables_compaction_task>(*this, t, std::move(all_sstables)));
 }
 
+future<> compaction_manager::maybe_wait_on_current_compaction(replica::table* t) {
+    auto sstables = boost::copy_range<std::vector<sstables::shared_sstable>>(
+            *(t->get_sstables()) |
+            boost::adaptors::filtered([](const sstables::shared_sstable& s) { return s->get_origin() == "memtable"; }));
+    sstables = get_non_compacting_sstables(sstables);
+    if (sstables.size() < t->schema()->max_compaction_threshold() || t->is_auto_compaction_disabled_by_user()) {
+        co_return;
+    }
+    auto tasks_compacting_table = boost::copy_range<std::vector<shared_ptr<task>>>(_tasks |
+            boost::adaptors::filtered([table=t](const shared_ptr<task>& t) { return t->compacting_table() == table; }));
+    if (tasks_compacting_table.empty()) {
+        co_return;
+    }
+
+    // Wait until any of the currently running compactions on the table finish.
+    auto s = seastar::make_lw_shared<std::optional<seastar::promise<>>>(seastar::promise<>());
+    for (auto t : tasks_compacting_table) {
+        (void)t->compaction_done().then([s] {
+            if (*s) {
+                (*s)->set_value();
+                *s = {};
+            }
+        });
+    }
+    co_await (*s)->get_future();
```

Comment:
A less tricky way to accomplish this is to use a condition variable. Any compacting task will signal it, and the code here can just loop around the condition variable checking the exit condition.

Reply:
We also want to exit if the server is shutting down (if we're under such stress it is likely the administrator will try to bring the server down). So need to make sure this wait can terminate (perhaps with an exception) if we're shutting down.

Reply:
With enough write rate, all tiers will be compacting in parallel. We shouldn't wait for compaction on larger tiers, but only on the smallest one. So it's indeed very important to wait for only one compaction, which is probably the one on smallest tier, not all of them. we can probably stop blocking flush if we know we're done with one compaction job at least. we don't need to start that hard, right?

Reply:
We should just wait until the number of sstables drops below the danger zone. We shouldn't care which compaction brought it down.

It can also go down due to a truncate, no?

Reply:
Ok, so it seems it doesn't have to go through compaction manager at all. we could have the sstable set update procedure(s) in table signaling the condition variable, then allowing the wait procedure to recheck the amount of L0 (with memtable origin) runs. It doesn't matter if the files are being compacted or not, we want to watch sstable set and wait until number of files drop.

Reply:
We could end up with a dead node. Although, forgetting auto compaction disabled will end up with a number of non-tiny L0 sstables, not the tiny sstables that are a result of schema flushes.

Reply:
Makes me think we need to log the blocking and possibly have a gauge metric too.